File: src/core/wallet.ts
// import { MPP } from '@monad-crypto/mpp';
import * as crypto from 'crypto';
import * as fs from 'fs';
import * as path from 'path';

class Wallet {
  private privateKey: string;
  private publicKey: string;
  private address: string;
  // private mpp: MPP;

  constructor(privateKey?: string) {
    if (privateKey) {
      this.privateKey = privateKey;
      this.publicKey = this.generatePublicKey(privateKey);
      this.address = this.generateAddress(this.publicKey);
    } else {
      const keys = this.generateKeyPair();
      this.privateKey = keys.privateKey;
      this.publicKey = keys.publicKey;
      this.address = this.generateAddress(keys.publicKey);
    }
    
    // this.mpp = new MPP({
    //   network: 'testnet',
    //   privateKey: this.privateKey
    // });
  }

  private generateKeyPair(): { privateKey: string; publicKey: string } {
    const { privateKey, publicKey } = crypto.generateKeyPairSync('ec', {
      namedCurve: 'secp256k1',
      publicKeyEncoding: { type: 'spki', format: 'pem' },
      privateKeyEncoding: { type: 'pkcs8', format: 'pem' }
    });
    return { privateKey, publicKey };
  }

  private generatePublicKey(privateKey: string): string {
    const keyObject = crypto.createPrivateKey(privateKey);
    return keyObject.export({ type: 'spki', format: 'pem' }).toString();
  }

  private generateAddress(publicKey: string): string {
    const publicKeyObject = crypto.createPublicKey(publicKey);
    const publicKeyBuffer = Buffer.from(publicKeyObject.export({ type: 'spki', format: 'der' }));
    const hash = crypto.createHash('sha256').update(publicKeyBuffer).digest();
    const address = '0x' + hash.slice(-20).toString('hex');
    return address;
  }

  getAddress(): string {
    return this.address;
  }

  getPublicKey(): string {
    return this.publicKey;
  }

  async getBalance(): Promise<number> {
    try {
      // 模拟余额
      return 1000;
    } catch (error) {
      console.error('Error getting balance:', error);
      return 0;
    }
  }

  async sendTransaction(to: string, amount: number): Promise<string> {
    try {
      // 模拟交易哈希
      return `0x${crypto.randomBytes(32).toString('hex')}`;
    } catch (error) {
      console.error('Error sending transaction:', error);
      throw error;
    }
  }

  saveToFile(filePath: string): void {
    const walletData = {
      privateKey: this.privateKey,
      publicKey: this.publicKey,
      address: this.address
    };
    fs.writeFileSync(filePath, JSON.stringify(walletData, null, 2));
  }

  static loadFromFile(filePath: string): Wallet {
    if (!fs.existsSync(filePath)) {
      throw new Error('Wallet file not found');
    }
    const walletData = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    return new Wallet(walletData.privateKey);
  }
}

export default Wallet;

File: app/core/wallet.py
import json
import os
from web3 import Web3
from eth_account import Account

class Wallet:
    def __init__(self, private_key: str = None):
        if private_key:
            if not private_key.startswith('0x'):
                private_key = '0x' + private_key
            self.account = Account.from_key(private_key)
        else:
            self.account = Account.create()
            
    def get_address(self) -> str:
        return self.account.address

    def get_private_key(self) -> str:
        return self.account.key.hex()

    def get_balance(self, rpc_url: str = "https://testnet-rpc.monad.xyz") -> int:
        try:
            w3 = Web3(Web3.HTTPProvider(rpc_url))
            balance = w3.eth.get_balance(self.account.address)
            return balance
        except Exception as e:
            print(f"Error getting balance: {e}")
            return 0

    def save_to_file(self, file_path: str) -> None:
        wallet_data = {
            "privateKey": self.get_private_key(),
            "address": self.get_address()
        }
        os.makedirs(os.path.dirname(file_path), exist_ok=True)
        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(wallet_data, f, indent=2)

    @classmethod
    def load_from_file(cls, file_path: str) -> 'Wallet':
        if not os.path.exists(file_path):
            raise FileNotFoundError("Wallet file not found")
            
        with open(file_path, 'r', encoding='utf-8') as f:
            wallet_data = json.load(f)
            
        private_key = wallet_data.get("privateKey")
        if not private_key:
            raise ValueError("Invalid wallet file: missing privateKey")
            
        return cls(private_key)
